Feature/create breeding

HenrikeW opened the pull request:
**UPDATE**
- I moved the functionality to create and update breedings/ births into a new, separate context file. 
This way it can be used both in the breeding form and in the individual_add component. 
Much less duplicated code now. 
Also, the linter is happy now :)
-----------------------------------------------------------------------------------------------------------------------------------------------------------------
- most of the layout for creating a breeding event was already there, I just changed the UI a bit to make it hopefully more user-friendly
- there was no functionality in place in the frontend to create or update breeding events, so I added this
- I added validation for the user input to not send API requests that are doomed to fail. 
- I also added a validation to check if there are about 30 days (28-32) between birth and breeding date. If you think it's unnecessary in the frontend, I can remove it.
- I created the API requests so that they would match the available endpoints (`/api/breeding`, `api/breeding/${herdId} and `api/birth`) as they look right now, even though I don't really see the point with creating breeding and birth separately. Why not put it into one endpoint?
- I also added two lines in the backend because the birth_notes and breeding_notes weren't saved. Copy-pasted the lines from another function but it seems to work. Due to my lack of Python skills, it'd be nice if you could have an extra look here :)

HenrikeW committed:
add createBreeding function to breeding_form

also added spara-button to call it.
still problems with the api-request.

HenrikeW committed:
improve ui for adding birth event

HenrikeW committed:
change breed_date to date

HenrikeW committed:
breeding data mandatory and autofill function

also added todo-comments

HenrikeW committed:
prepare to add birth information

move create-breeding-part into separate function.
create interfaces for birth / breeding information.
rename functions.

HenrikeW committed:
add createBirth function to save birth information

HenrikeW committed:
make date fields clearable

HenrikeW committed:
add function updateBreeding...

also rename parameter in saveBreeding.
Add explaining comments.
Reorder functions within the file.

HenrikeW committed:
remove unused variables

HenrikeW committed:
disable future dates

HenrikeW committed:
add function to validate date input

HenrikeW committed:
improve litter validation

HenrikeW committed:
make date errors less aggressive

HenrikeW committed:
removing bugs from update breeding functionality

rename date to breed_date.
don't send empty object keys to backend.
allow litter size to be null.
save breed notes and birth notes in backend.

HenrikeW committed:
move user input validation into separate function

HenrikeW committed:
rename variables and restructure order in breeding_form

HenrikeW committed:
fix validation function and make date fields clearable

HenrikeW committed:
make litter size clearable

HenrikeW committed:
remove some TS errors

HenrikeW committed:
let breeding updates appear in table

HenrikeW committed:
always clear form on addbutton click

jonandernovella commented on frontend/src/breeding_form.tsx:
```diff
@@ -116,22 +129,311 @@ export function BreedingForm({
+      return;
+    }
+
+    const currentHerdBreedings = await get(`/api/breeding/${herdId}`);
```
This could be in a different tsx file like we have in the backend the data_access layer.

jonandernovella replied:
The reason for this is that we can reuse the code for this api calls and data type, which is used also in other components.

jonandernovella commented on frontend/src/breeding_form.tsx:
```diff
@@ -116,22 +129,311 @@ export function BreedingForm({
+      breeding_id?: number;
+    } = await post("/api/breeding", breedingData);
+
+    if (breedingCreationResponse.status == "success") {
+      return breedingCreationResponse;
+    }
```
this looks like a good workaround as long as we don't get proper http codes back

jonandernovella reviewed (approved):
Good work. I left a few picky comments :)

jonandernovella commented:
Check the linter. There seems to be some complaint!

HenrikeW committed:
making breeding functions reusable (in progress)

HenrikeW committed:
breeding form works with imported functions

HenrikeW committed:
get everything into centext file

restructured individual_add.
changed context order in main.

HenrikeW committed:
debugging. And reordering individual_add

HenrikeW committed:
add check for herd/number match in individual_add

HenrikeW committed:
add check for litter size in individual_add

HenrikeW committed:
add comments to individual_add

HenrikeW committed:
move breeding functionality to new context file

this is cleaner to hold data context and breeding context apart.
also remove await from modifyBreedingUpdates as it is not asynchronous.

HenrikeW committed:
make translate map global to make linter happy

HenrikeW committed:
Merge pull request #347 from NBISweden/dev/context_functions

Dev/context functions

jonandernovella commented on frontend/src/breeding_context.tsx:
```diff
@@ -0,0 +1,204 @@
+   * This way, the right breeding is updated and no empty fields are sent to the API.
+   * Empty fields would cause error messages from the API.
+   * */
+  const modifyBreedingUpdates = (updates: Breeding, breedingMatch) => {
```
Why is breeding match not typed here?

HenrikeW replied:
I don't know, actually. Added it.

jonandernovella commented on frontend/src/breeding_form.tsx:
```diff
@@ -57,21 +64,30 @@ const emptyBreeding: Breeding = {
 }: {
   data: Breeding | "new";
   herdId: string | undefined;
+  handleBreedingsChanged: any;
+  handleActive: any;
```
Why any? They sound like a boolean.

HenrikeW replied:
Nope, they are functions. But not "any" anyway, so I changed it. Thanks!

jonandernovella commented on frontend/src/breeding_form.tsx:
```diff
@@ -41,13 +48,13 @@ const useStyles = makeStyles({
+  birth_date: null,
   birth_notes: "",
-  litter_size: 0,
+  litter_size: null,
```
What about 0?

HenrikeW replied:
I want the input field empty when the page renders. It's only empty if the value is null, otherwise it shows a 0.

jonandernovella commented on frontend/src/main.tsx:
```diff
@@ -32,7 +33,9 @@ const Main = (
         <WithUserContext>
           <WithMessageContext>
-            <Navigation />
+            <WithBreedingContext>
+              <Navigation />
+            </WithBreedingContext>
```
Neat!

jonandernovella reviewed (commented):
Left a few thoughts. Otherwise it looks in good shape!

HenrikeW committed:
update some types due to Jon's comments

HenrikeW commented:
Thanks for having eagle eyes, Jon! Added some types where they were wrong / missing but would like to keep it null instead of 0, see above. If you're happy with it, feel free to merge.
